language-plutus-core/src/Language/PlutusCore/Constant/Pretty.hs

```diff
@@ -0,0 +1,74 @@
+{-# OPTIONS_GHC -fno-warn-orphans #-}
```

Comment:
Why the orpans? It may clutter up modules but if we do not introduce any orphans we will not run into subtle conflicts later on.

Reply:
I think this is a relatively benign case of orphans, since we're unlikely to be defining complex instances anyway.

Reply:
Having a separate `Pretty` module is still problematic, IMO.

Reply:
> Hackage is full of orphan instances and generally they're unavoidable in the current Haskell. How are module boundaries different from package boundaries?

Yes, in part because such things are unavoidable without upstream pull requests. They are not *necessary* across module boundaries.

> not preventing a problem that is very unlikely to pop up.

I do not believe it is unlikely to pop up, and in any case, it would be difficult to debug when it did.

> GHC.Base has -fno-warn-orphans.

`base` uses orphan instances less liberally.

> That doesn't seem like a situation that we have to prevent

It would be frustrating to debug if it did.

 >Those are perfectly separated concepts, "-Wall said this" is not a good reason to make spaghetti out of code.

There are some warnings more worthwhile than others, and warnings about orphans are more relevant. It is not a question of separated concepts but rather avoiding overlapping instances by accident.

Reply:
> Yes, in part because such things are unavoidable without upstream pull requests. They are not *necessary* across module boundaries.

Even with zero-cost upstream pull requests, there still would be orphan instances, because if you have a package `X` that defines a type `D` and a package `Y` that defines a class `C`

1. it is not always the case that there is some natural ordering of what is "more important", i.e. whether to import `X` from `Y` or `Y` from `X`
2. you do want to keep dependencies small to avoid putting maintenance burden on the shoulders of a single party (this doesn't apply as much to the modules case, but is still relevant) and to preserve encapsulation: each package/entity solves a particular problem, others build on top of that (this applies to our case as much)

> I do not believe it is unlikely to pop up

I worked with a guy that maintains more than a hundred packages on Hackage and does not care about orphan instances at all and I've never heard him saying "oh, I got a problem due to that". I would be really curious to hear a story of orphans causing problems in practice rather than theory. Especially when everything involved is defined in the same package.

> It is not a question of separated concepts but rather avoiding overlapping instances by accident.

It is a question of both and I believe the former is more important, because defining an instance twice in the same package and not noticing it and using different instances in different scenarios is highly unlikely.

Here is what Kmett thinks:

> In theory, they are terrible. ;)

> In practice sometimes they are unavoidable, as later libraries added to hackage that expose properties of a large number of built-in types have to choose between providing instances for everything and the kitchen sink out of pre-existing libraries that are out there, thereby bloating their import list and limiting the number of scenarios in which they are applicable, or provide a series of packages which exist solely to provide orphan instances, and thereby risk accidental non-inclusion and the host of problems that orphan instances can entail.

I removed all orphan instances, because I believe this conversation does more harm to my productivity than this change to our codebase, but it made `Language.PlutusCore.Name` exactly the kitchen sink from the quote above.